src/test/metrics_test.cpp

```diff
@@ -0,0 +1,780 @@
+    EXPECT_THAT(registry.collect(), HasSubstr("fam{other=\"data\"} 0"));
+    EXPECT_TRUE(registry.remove(family1));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("# HELP name")));
+    EXPECT_THAT(registry.collect(), HasSubstr("# HELP fam"));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("# TYPE name")));
+    EXPECT_THAT(registry.collect(), HasSubstr("# TYPE fam"));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name{label=\"value\"}")));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name{other=\"data\"}")));
+    EXPECT_THAT(registry.collect(), HasSubstr("fam{other=\"data\"} 0\n"));
+}
+
+TEST(MetricsGauge, RemoveRemovedFamily) {
+    MetricRegistry registry;
+    auto family = registry.createFamily<MetricGauge>("name", "desc");
+    family->addMetric();
+    EXPECT_TRUE(registry.remove(family));
+    EXPECT_FALSE(registry.remove(family));
+}
+
+TEST(MetricsGauge, RevertingMetricResetsValue) {
+    MetricRegistry registry;
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name{label=\"value\"}")));
+    auto family = registry.createFamily<MetricGauge>("name", "desc");
+    auto metric = family->addMetric({{"label", "value"}});
+    EXPECT_THAT(registry.collect(), HasSubstr("name{label=\"value\"} 0\n"));
+    metric->increment();
+    metric->decrement();
+    metric->decrement();
+    EXPECT_THAT(registry.collect(), HasSubstr("name{label=\"value\"} -1\n"));
+    family->remove(metric);
+    metric = family->addMetric({{"label", "value"}});
+    EXPECT_THAT(registry.collect(), HasSubstr("name{label=\"value\"} 0\n"));
+}
+
+TEST(MetricsGauge, RevertingFamilyResetsValue) {
+    MetricRegistry registry;
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name{label=\"value\"}")));
+    auto family = registry.createFamily<MetricGauge>("name", "desc");
+    auto metric = family->addMetric({{"label", "value"}});
+    EXPECT_THAT(registry.collect(), HasSubstr("name{label=\"value\"} 0\n"));
+    metric->increment();
+    metric->decrement();
+    metric->decrement();
+    EXPECT_THAT(registry.collect(), HasSubstr("name{label=\"value\"} -1\n"));
+    registry.remove(family);
+    family = registry.createFamily<MetricGauge>("name", "desc");
+    metric = family->addMetric({{"label", "value"}});
+    EXPECT_THAT(registry.collect(), HasSubstr("name{label=\"value\"} 0\n"));
+}
+
+TEST(MetricsHistogram, Observe) {
+    MetricRegistry registry;
+    auto metric = registry.createFamily<MetricHistogram>("name", "desc")->addMetric({{"label", "value"}}, {1.0, 10.0});
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{label=\"value\",le=\"1\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{label=\"value\",le=\"10\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{label=\"value\",le=\"+Inf\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_count{label=\"value\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_sum{label=\"value\"} 0\n"));
```

Comment:
Not sure if checking histogram buckets when they are just created is necessary as well.

Reply:
Removed